underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AUnderFileSystem.java

```diff
@@ -0,0 +1,693 @@
+    if (details != null) {
+      return details.getLastModified().getTime();
+    } else {
+      throw new FileNotFoundException(path);
+    }
+  }
+
+  // This call is currently only used for the web ui, where a negative value implies unknown.
+  @Override
+  public long getSpace(String path, SpaceType type) throws IOException {
+    return -1;
+  }
+
+  @Override
+  public boolean isFile(String path) throws IOException {
+    return exists(path) && !isFolder(path);
+  }
+
+  @Override
+  public String[] list(String path) throws IOException {
+    // if the path not exists, or it is a file, then should return null
+    if (!exists(path) || isFile(path)) {
+      return null;
+    }
+    // Non recursive list
+    path = PathUtils.normalizePath(path, PATH_SEPARATOR);
+    return listInternal(path, false);
+  }
+
+  @Override
+  public boolean mkdirs(String path, boolean createParent) throws IOException {
+    return mkdirs(path, new MkdirsOptions().setCreateParent(createParent));
+  }
+
+  @Override
+  public boolean mkdirs(String path, MkdirsOptions options) throws IOException {
+    if (path == null) {
+      return false;
+    }
+    if (isFolder(path)) {
+      return true;
+    }
+    if (exists(path)) {
+      LOG.error("Cannot create directory {} because it is already a file.", path);
+      return false;
+    }
+    if (!options.getCreateParent()) {
+      if (parentExists(path)) {
+        // Parent directory exists
+        return mkdirsInternal(path);
+      } else {
+        LOG.error("Cannot create directory {} because parent does not exist", path);
+        return false;
+      }
+    }
+    // Parent directories should be created
+    if (parentExists(path)) {
+      // Parent directory exists
+      return mkdirsInternal(path);
+    } else {
+      String parentKey = getParentKey(path);
+      // Recursively make the parent folders
+      return mkdirs(parentKey, true) && mkdirsInternal(path);
+    }
+  }
+
+  @Override
+  public InputStream open(String path) throws IOException {
+    try {
+      path = stripPrefixIfPresent(path);
+      return new S3AInputStream(mBucketName, path, mClient);
+    } catch (AmazonClientException e) {
+      LOG.error("Failed to open file: {}", path, e);
+      return null;
+    }
+  }
+
+  /**
+   * Opens a S3 object at given position and returns the opened input stream.
+   *
+   * @param path the S3 object path
+   * @param pos the position to open at
+   * @return the opened input stream
+   * @throws java.io.IOException if failed to open file at position
+   */
+  public InputStream openAtPosition(String path, long pos) throws IOException {
+    try {
+      path = stripPrefixIfPresent(path);
+      return new S3AInputStream(mBucketName, path, mClient, pos);
+    } catch (AmazonClientException e) {
+      LOG.error("Failed to open file {} at position {}:", path, pos, e);
+      return null;
+    }
+  }
+
+  @Override
+  public boolean rename(String src, String dst) throws IOException {
```

Comment:
Good point. Let's change the rename semantics for all the under storages after we decide on something.